huaweicloud/services/workspace/resource_huaweicloud_workspace_desktop.go

```diff
@@ -607,6 +629,81 @@ func updateDesktopNetwork(ctx context.Context, client *golangsdk.ServiceClient,
 	return nil
 }
 
+func waitForWorkspaceStatusCompleted(ctx context.Context, client *golangsdk.ServiceClient, d *schema.ResourceData) error {
+	stateConf := &resource.StateChangeConf{
+		Pending:      []string{"PENDING"},
+		Target:       []string{"COMPLETED"},
+		Refresh:      desktopStatusRefreshFunc(client, d),
+		Timeout:      d.Timeout(schema.TimeoutUpdate),
+		Delay:        10 * time.Second,
+		PollInterval: 10 * time.Second,
+		// The final status of both startup and restart is "ACTIVE". the parameter only applies to "REBOOT" action.
+		ContinuousTargetOccurence: 2,
+	}
+	_, err := stateConf.WaitForStateContext(ctx)
+	return err
+}
+
+func desktopStatusRefreshFunc(client *golangsdk.ServiceClient, d *schema.ResourceData) resource.StateRefreshFunc {
+	return func() (interface{}, string, error) {
+		desktopId := d.Id()
+		resp, err := desktops.Get(client, desktopId)
+		if err != nil {
+			return resp, "", err
+		}
+
+		taskStatus := resp.TaskStatus
+		// In statusMap, key represents the power action of the desktop, and value represents the status after the desktop operation is completed.
+		// Use statusMap to make a mapping relationship between the power action of the desktop and the final status of the desktop.
+		statusMap := map[string]string{
+			"os-start":     "ACTIVE",
+			"os-stop":      "SHUTOFF",
+			"reboot":       "ACTIVE",
+			"os-hibernate": "SHUTOFF",
+		}
+
+		powerAction := d.Get("power_action").(string)
+		// TaskStatus variable is always an empty string when the desktop power action is completed.
+		// If the desktop power action changes from one state to another, taskStatus is an empty string for a long time,
+		// whether a desktop action is completed cannot be determined only by taskStatus.
+		if taskStatus == "" && resp.Status == statusMap[powerAction] {
```

Comment:
please define taskStatus as near as it was used, or using `resp.TaskStatus` directly
```suggestion
		if resp.TaskStatus == "" && resp.Status == statusMap[powerAction] {
```

Reply:
fixed